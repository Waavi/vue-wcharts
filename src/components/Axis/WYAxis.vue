<template>
    <g id="WYAxis">
        <line
            v-if="!hideLine"
            :x1="x1"
            :y1="y1"
            :x2="x2"
            :y2="y2"
            :stroke="axisStylesCmp.stroke"
        />
        <template
            v-for="(tick, index) in ticks"
        >
            <g
                :key="`tick-${index}`"
                text-anchor="end"
            >
                <line
                    v-if="!hideTickMark"
                    :x1="tick.mark.x1"
                    :y1="tick.mark.y1"
                    :x2="tick.mark.x2"
                    :y2="tick.mark.y2"
                    :stroke="markStylesCmp.stroke"
                />
                <slot
                    name="tick"
                    :tick="tick.text"
                >
                    <Tick
                        :text="tick.text"
                        :styles="labelStylesCmp"
                    />
                </slot>
            </g>
        </template>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import Tick from './Tick.vue'
import { Maths } from '../../utils'

const axisStylesDefaultProp = {
    stroke: '#999',
}

const markStylesDefaultProp = {
    stroke: '#999',
}

const labelStylesDefaultProp = {
    stroke: 'none',
    fill: '#999',
    fontSize: 12,
}

export default {
    name: 'WYAxis',
    type: 'axis',
    inject: ['Cartesian'],
    components: {
        Tick,
    },
    props: {
        space: VueTypes.arrayOf(VueTypes.number).def([10, 0, 0, 40]),
        textOffsetY: VueTypes.number.def(10),
        ticksNum: VueTypes.number,
        hideLine: VueTypes.bool.def(false),
        hideTickMark: VueTypes.bool.def(false),
        /** Styles */
        axisStyles: VueTypes.shape({
            stroke: VueTypes.string,
        }).def(() => ({
            ...axisStylesDefaultProp,
        })),
        markStyles: VueTypes.shape({
            stroke: VueTypes.string,
        }).def(() => ({
            ...axisStylesDefaultProp,
        })),
        labelStyles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            fontSize: VueTypes.oneOfType([String, Number]),
        }).def(() => ({
            ...labelStylesDefaultProp,
        })),
    },
    computed: {
        ticks () {
            const {
                dataset, canvas, bounds, padding,
            } = this.Cartesian
            const num = this.ticksNum || dataset.length
            const ticks = Maths.genTicks(bounds.min, bounds.max, num).reverse()
            const offset = (padding[0] + padding[2])
            const space = (canvas.height - offset) / (ticks.length - 1)

            return ticks.map((value, index) => {
                const y = canvas.y0 + (space * index) + padding[2]
                return {
                    mark: {
                        index,
                        x1: canvas.x0 - 5,
                        y1: y,
                        x2: canvas.x0,
                        y2: y,
                    },
                    text: {
                        index,
                        value,
                        x: canvas.x0 - this.textOffsetY,
                        y: y + this.labelStylesCmp.fontSize / 3,
                    },
                }
            })
        },
        x1 () {
            return this.Cartesian.canvas.x0
        },
        y1 () {
            return this.Cartesian.canvas.y0
        },
        x2 () {
            return this.Cartesian.canvas.x0
        },
        y2 () {
            return this.Cartesian.canvas.y1
        },
        axisStylesCmp () {
            return {
                ...axisStylesDefaultProp,
                ...this.axisStyles,
            }
        },
        markStylesCmp () {
            return {
                ...markStylesDefaultProp,
                ...this.markStyles,
            }
        },
        labelStylesCmp () {
            return {
                ...labelStylesDefaultProp,
                ...this.labelStyles,
            }
        },
    },
}
</script>
