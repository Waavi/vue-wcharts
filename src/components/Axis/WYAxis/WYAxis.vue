<template>
    <g :id="id">
        <line
            v-if="!hideLine"
            :x1="x1"
            :y1="y1"
            :x2="x2"
            :y2="y2"
            :stroke="axisStylesCmp.stroke"
        />
        <template
            v-for="(tick, index) in ticks"
        >
            <g
                :key="`tick-${index}`"
                :text-anchor="textAnchor"
            >
                <line
                    v-if="!hideTickMark"
                    :x1="tick.mark.x1"
                    :y1="tick.mark.y1"
                    :x2="tick.mark.x2"
                    :y2="tick.mark.y2"
                    :stroke="markStylesCmp.stroke"
                />
                <slot
                    name="tick"
                    v-bind="tick.text"
                    :styles="tickStylesCmp"
                >
                    <WAxisText
                        v-bind="tick.text"
                        :styles="tickStylesCmp"
                    >
                        <template #default="{ value }">
                            <slot
                                name="tickText"
                                :value="value"
                            >
                                {{ value }}
                            </slot>
                        </template>
                    </WAxisText>
                </slot>
            </g>
        </template>

        <slot
            name="label"
            v-bind="labels"
            :styles="labelStylesCmp"
        >
            <WAxisLabel
                v-if="label"
                v-bind="labels"
                :styles="labelStylesCmp"
            />
        </slot>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import axisMixin from '../../../mixins/axis'

export default {
    name: 'WYAxis',
    axis: 'y',
    mixins: [axisMixin],
    props: {
        labelAlign: VueTypes.oneOf(['start', 'middle', 'end']).def('start'),
        space: VueTypes.arrayOf(VueTypes.number).def([10, 0, 0, 40]),
        textOffset: VueTypes.number.def(10),
    },
}
</script>
