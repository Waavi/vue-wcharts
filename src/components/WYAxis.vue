<template>
    <g stroke="#999">
        <line
            :x1="x1"
            :y1="y1"
            :x2="x2"
            :y2="y2"
        />
    </g>
</template>

<script>
import VueTypes from 'vue-types'

export default {
    name: 'WYAxis',
    type: 'axis',
    inject: ['Cartesian'],
    props: {
        datakey: VueTypes.string,
        space: VueTypes.arrayOf(VueTypes.number).def([10, 0, 0, 40]),
    },
    computed: {
        points () {
            return this.Cartesian.dataset.map(value => value.name)
        },
        x1 () {
            return this.Cartesian.canvas.x0
        },
        y1 () {
            return this.Cartesian.canvas.y0
        },
        x2 () {
            return this.Cartesian.canvas.x0
        },
        y2 () {
            return this.Cartesian.canvas.y1
        },
    },
}
</script>
