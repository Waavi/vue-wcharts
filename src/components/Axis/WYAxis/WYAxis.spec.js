import { shallowMount } from '@vue/test-utils'
import WYAxis from './WYAxis.vue'

describe('Axis/WYAxis', () => {
    const dataset = [
        {
            name: 'Page A',
            one: 1000,
            two: 2400,
            three: 2400,
        },
        {
            name: 'Page B',
            one: 3000,
            two: 1398,
            three: 2210,
        },
        {
            name: 'Page C',
            one: 2000,
            two: 9800,
            three: 2290,
        },
    ]

    const canvas = {
        x0: 40,
        y0: 10,
        width: 540,
        height: 366,
        x1: 580,
        y1: 376,
    }

    const bounds = {
        max: 8800,
        min: 10,
    }

    const propsData = {
        label: 'Y Axis',
    }

    const provide = {
        Chart: {
            dataset,
            data: dataset,
            canvas,
            bounds,
            spaceObjects: [10, 10, 10, 10],
            yScale: a => a,
            setAxisOptions: () => undefined,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WYAxis, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly without line`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, hideLine: true } }
        const wrapper = shallowMount(WYAxis, customConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly without ticks lines`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, hideTickMark: true } }
        const wrapper = shallowMount(WYAxis, customConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })
})
