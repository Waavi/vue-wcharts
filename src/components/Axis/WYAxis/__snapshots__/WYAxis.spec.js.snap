// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Axis/WYAxis Should be render correctly 1`] = `
<g id="WYAxis">
  <line x1="40" y1="10" x2="40" y2="376" stroke="#cccccc"></line>
  <g text-anchor="end">
    <line x1="35" y1="8800" x2="40" y2="8800" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="8804" dy="" index="0" value="8800" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="7544" x2="40" y2="7544" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="7548" dy="" index="1" value="7544" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="6288" x2="40" y2="6288" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="6292" dy="" index="2" value="6288" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="5032" x2="40" y2="5032" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="5036" dy="" index="3" value="5032" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="3777" x2="40" y2="3777" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="3781" dy="" index="4" value="3777" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="2521" x2="40" y2="2521" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="2525" dy="" index="5" value="2521" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="1265" x2="40" y2="1265" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="1269" dy="" index="6" value="1265" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="10" x2="40" y2="10" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="14" dy="" index="7" value="10" styles="[object Object]"></WAxisText-stub>
  </g>
  <WAxisLabel styles="[object Object]" x="17" y="376" transform="rotate(-90 17 376)" textAnchor="start" value="Y Axis"></WAxisLabel>
</g>
`;

exports[`Axis/WYAxis Should be render correctly without line 1`] = `
<g id="WYAxis">
  <!---->
  <g text-anchor="end">
    <line x1="35" y1="8800" x2="40" y2="8800" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="8804" dy="" index="0" value="8800" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="7544" x2="40" y2="7544" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="7548" dy="" index="1" value="7544" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="6288" x2="40" y2="6288" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="6292" dy="" index="2" value="6288" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="5032" x2="40" y2="5032" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="5036" dy="" index="3" value="5032" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="3777" x2="40" y2="3777" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="3781" dy="" index="4" value="3777" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="2521" x2="40" y2="2521" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="2525" dy="" index="5" value="2521" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="1265" x2="40" y2="1265" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="1269" dy="" index="6" value="1265" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <line x1="35" y1="10" x2="40" y2="10" stroke="#cccccc"></line>
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="14" dy="" index="7" value="10" styles="[object Object]"></WAxisText-stub>
  </g>
  <WAxisLabel styles="[object Object]" x="17" y="376" transform="rotate(-90 17 376)" textAnchor="start" value="Y Axis"></WAxisLabel>
</g>
`;

exports[`Axis/WYAxis Should be render correctly without ticks lines 1`] = `
<g id="WYAxis">
  <line x1="40" y1="10" x2="40" y2="376" stroke="#cccccc"></line>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="8804" dy="" index="0" value="8800" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="7548" dy="" index="1" value="7544" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="6292" dy="" index="2" value="6288" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="5036" dy="" index="3" value="5032" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="3781" dy="" index="4" value="3777" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="2525" dy="" index="5" value="2521" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="1269" dy="" index="6" value="1265" styles="[object Object]"></WAxisText-stub>
  </g>
  <g text-anchor="end">
    <!---->
    <WAxisText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" x="30" y="14" dy="" index="7" value="10" styles="[object Object]"></WAxisText-stub>
  </g>
  <WAxisLabel styles="[object Object]" x="17" y="376" transform="rotate(-90 17 376)" textAnchor="start" value="Y Axis"></WAxisLabel>
</g>
`;
